nupic/research/frameworks/dendrites/routing/routing.py

```diff
@@ -19,13 +19,94 @@
 # http://numenta.org/licenses/
 # ----------------------------------------------------------------------
 
+from random import randint
+
 import torch
+from torch.utils.data import Dataset
 
 from nupic.torch.modules import SparseWeights
 
 from .utils import generate_random_binary_vectors
 
 
+class RoutingDataset(Dataset):
+    """
+    A dataset class for generating input-target pairs for the routing test specifically
+    for a linear network with a single layer, where the inputs are random vectors
+    sampled from U[-2, 2) each paired a binary sparse context vector
+    """
+
+    def __init__(
+        self,
+        routing_function,
+        input_size,
+        context_vectors,
+        device,
+        concat=False,
+        dataset_size=1e4,
+        x_min=-2.0,
+        x_max=2.0,
+    ):
+        """
+        :param routing_function: the random routing function
+        :param input_size: the number of dimensions in the input to the routing
+                           function and test module
+        :param context_vectors: 2D torch Tensor in which each row gives a context
+                                vector
+        :param device: device to use ('cpu' or 'cuda')
+        :param concat: if True, input and context vectors are concatenated together
+        :param dataset_size: the number of (input, context, target) pairs that be
+                             iterated over
+        :param x_min: the minimum bound of the uniform distribution from which input
+                      vectors are i.i.d. sampled along each input dimension
+        :param x_max: the maximum bound of the uniform distribution from which input
+                      vectors are i.i.d. sampled along each input dimension
+        """
+        super().__init__()
+        self.function = routing_function
+        self.num_output_masks = routing_function.num_output_masks
+        self.input_size = input_size
+        self.context_vectors = context_vectors
+        self.device = device
+        self.concat = concat
+        self.size = int(dataset_size)
+
+        # The following attributes are selected such that self.alpha * u + self.beta
+        # gives a sample drawn from U[x_min, x_max) given a sample u ~ U[0, 1)
+        self.alpha = (x_max - x_min)
+        self.beta = x_min
+
+    def __getitem__(self, idx):
+
+        # To retrieve an input-context-target pair, first generate noise from a uniform
+        # distribution as input (where the bounds of the distribution were specified in
+        # the `__init__` method), and take the routing function's output on said input
+        # using any of its output masks
+
+        if idx > self.size:
+            raise IndexError("Index {} is out of range".format(idx))
+        torch.manual_seed(idx)
+
+        x = self.alpha * torch.rand((self.input_size,)) - self.beta
```

Comment:
Yes, I will fix this. As an aside, I don't think this will make a big difference in terms of results, since right now we're sampling inputs i.i.d. from U(2, 6) instead of U(-2, 2).